sdk/core/azure-core/src/main/java/com/azure/core/http/policy/DefaultRedirectStrategy.java

```diff
@@ -0,0 +1,199 @@
+            tryGetRedirectHeader(httpResponse.getHeaders(), this.getLocationHeader());
+        if (responseLocation != null) {
+            attemptedRedirectUrls.add(responseLocation);
+            return httpResponse.getRequest().setUrl(responseLocation);
+        } else {
+            return httpResponse.getRequest();
+        }
+    }
+
+    @Override
+    public int getMaxAttempts() {
+        return maxAttempts;
+    }
+
+    @Override
+    public String getLocationHeader() {
+        return locationHeader;
+    }
+
+    @Override
+    public Map<Integer, HttpMethod> getAllowedMethods() {
+        return redirectMethods;
+    }
+
+    /**
+     * Check if the redirect url provided in the response headers is already attempted.
+     *
+     * @param redirectUrl the redirect url provided in the response header.
+     * @return {@code true} if the redirectUrl provided in the response header is already being attempted for redirect.
+     */
+    private boolean alreadyAttemptedRedirectUrl(String redirectUrl) {
+        if (attemptedRedirectUrls.contains(redirectUrl)) {
+            logger.error(String.format("Request was redirected more than once to: %s", redirectUrl));
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Check if the attempt count of the redirect is less than the {@code maxAttempts}
+     *
+     * @param tryCount the try count for the HTTP request associated to the HTTP response.
+     * @return {@code true} if the {@code tryCount} is greater than the {@code maxAttempts}.
+     */
+    private boolean isValidRedirectCount(int tryCount) {
+        if (tryCount >= getMaxAttempts()) {
+            logger.error(String.format("Request has been redirected more than %d times.", getMaxAttempts()));
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Check if the request http method is a valid redirect method.
+     *
+     * @param httpMethod the http method of the request.
+     * @return {@code true} if the request {@code httpMethod} is a valid http redirect method.
+     */
+    private boolean isAllowedRedirectMethod(HttpMethod httpMethod) {
+        if (getAllowedMethods().containsKey(httpMethod.ordinal())) {
+            return true;
+        } else {
+            logger.error(
+                String.format("Request was redirected from a non redirect-able method: %s", httpMethod));
+            return false;
+        }
+    }
+
+    /**
+     * Checks if the incoming request status code is a valid redirect status code.
+     *
+     * @param statusCode the status code of the incoming request.
+     * @return {@code true} if the request {@code statusCode} is a valid http redirect method.
+     */
+    private boolean isValidRedirectStatusCode(int statusCode) {
+        return statusCode == HttpURLConnection.HTTP_MOVED_TEMP
+            || statusCode == HttpURLConnection.HTTP_MOVED_PERM
+            || statusCode == PERMANENT_REDIRECT_STATUS_CODE;
```

Comment:
We should include [temporary redirect](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/307) too.